pkg/kv/bulk/split_sst_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package bulk
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/storageutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSplitSST(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(ctx)
+
+	pointKV := storageutils.PointKV
+	first := pointKV("b", 7, "foo")
+	middle := pointKV("c", 6, "baz")
+	last := pointKV("d", 6, "bar")
+	kvs := storageutils.KVs{first, middle, last}
+	sstBytes, start, end := storageutils.MakeSST(t, s.ClusterSettings(), kvs)
+
+	sp := func(start, end roachpb.Key, keyCount int) sstSpan {
+		stats := enginepb.MVCCStats{KeyCount: int64(keyCount)}
+		return sstSpan{start: start, end: end, stats: stats}
+	}
+
+	iterOpts := storage.IterOptions{
+		KeyTypes:   storage.IterKeyTypePointsOnly,
+		LowerBound: start,
+		UpperBound: end.Next(),
+	}
+
+	type testCase struct {
+		name          string
+		splitPoint    roachpb.Key
+		errorContains string
+		expectedLeft  sstSpan
+		expectedRight sstSpan
+	}
+
+	for _, tc := range []testCase{
+		{
+			name:          "first",
+			splitPoint:    first.Key.Key,
+			errorContains: "original sst must be larger than split key",
+		},
+		{
+			name:          "middle",
+			splitPoint:    middle.Key.Key,
+			expectedLeft:  sp(first.Key.Key, first.Key.Key.Next(), 1),
+			expectedRight: sp(middle.Key.Key, last.Key.Key.Next(), 2),
+		},
+		{
+			name:          "last",
+			splitPoint:    last.Key.Key,
+			expectedLeft:  sp(first.Key.Key, middle.Key.Key.Next(), 2),
+			expectedRight: sp(last.Key.Key, last.Key.Key.Next(), 1),
+		},
+		{
+			name:          "lastNext",
+			splitPoint:    last.Key.Key.Next(),
+			errorContains: "after last key",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			iter, err := storage.NewMemSSTIterator(sstBytes, true, iterOpts)
+			require.NoError(t, err)
+			defer iter.Close()
+
+			iter.SeekGE(storage.MVCCKey{Key: start})
+			sendStart := timeutil.Now()
+			stats, err := storage.ComputeStatsForIter(iter, sendStart.UnixNano())
```

Comment:
one surprise here: `ComputStatsIter` eventually invalidates the iter, yet `createSplitSSTable` can reuse it by seeking to the start point again.